File: oracle/verifier/stripe.go
package verifier

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"net/http"
	"strconv"
	"strings"
	"time"
)

func init() {
	registerProvider("stripe", func(r Route, secret string, _ verifierDeps) (Verifier, error) {
		if secret == "" {
			return nil, errors.New("empty secret")
		}
		window, err := parseWindow(r.ReplayWindow)
		if err != nil {
			return nil, fmt.Errorf("replay_window: %w", err)
		}
		return StripeVerifier{Secret: []byte(secret), ReplayWindow: window}, nil
	})
}

// StripeVerifier implements the Stripe signature shape: a Stripe-Signature
// header of the form "t=<unix>,v1=<hex>[,v1=<hex>...]", where the HMAC-SHA256 is
// computed over "<t>.<rawBody>". A timestamp outside ReplayWindow is rejected
// even when the HMAC matches.
type StripeVerifier struct {
	Secret       []byte
	ReplayWindow time.Duration
}

func (v StripeVerifier) Verify(rawBody []byte, h http.Header, now time.Time) error {
	header := h.Get("Stripe-Signature")
	if header == "" {
		return errors.New("missing Stripe-Signature header")
	}
	ts, sigs := parseStripeSig(header)
	if ts == "" || len(sigs) == 0 {
		return errors.New("malformed Stripe-Signature header")
	}

	tsInt, err := strconv.ParseInt(ts, 10, 64)
	if err != nil {
		return errors.New("invalid timestamp")
	}
	if v.ReplayWindow > 0 {
		delta := now.Sub(time.Unix(tsInt, 0))
		if delta < 0 {
			delta = -delta
		}
		if delta > v.ReplayWindow {
			return errors.New("timestamp outside replay window")
		}
	}

	mac := hmac.New(sha256.New, v.Secret)
	mac.Write([]byte(ts))
	mac.Write([]byte("."))
	mac.Write(rawBody)
	expected := mac.Sum(nil)

	for _, s := range sigs {
		got, err := hex.DecodeString(s)
		if err != nil {
			continue
		}
		if hmac.Equal(got, expected) {
			return nil
		}
	}
	return errors.New("no matching signature")
}

func parseStripeSig(header string) (ts string, sigs []string) {
	for _, part := range strings.Split(header, ",") {
		kv := strings.SplitN(strings.TrimSpace(part), "=", 2)
		if len(kv) != 2 {
			continue
		}
		switch kv[0] {
		case "t":
			ts = kv[1]
		case "v1":
			sigs = append(sigs, kv[1])
		}
	}
	return ts, sigs
}

File: src/hookguard_gateway/providers/stripe.py
"""Stripe's signature shape.

A ``Stripe-Signature`` header of the form ``t=<unix>,v1=<hex>[,v1=<hex>...]``,
where the HMAC-SHA256 is computed over ``"<t>.<rawBody>"``. Stripe is the one
supported provider whose signature carries a timestamp, so it is the only one
with a replay window: a timestamp outside the window is rejected even when the
HMAC matches.
"""

from __future__ import annotations

import hashlib
import hmac
from dataclasses import dataclass
from datetime import datetime, timedelta

from hookguard_core.goduration import GoDurationError, parse_go_duration
from hookguard_core.stricthex import decode_hex

from ..config import Route
from ..verifier import VerificationError, Verifier, VerifierDeps, register_provider

__all__ = ["StripeVerifier"]

HEADER = "Stripe-Signature"


@register_provider("stripe")
def _build(route: Route, secret: str, _deps: VerifierDeps) -> Verifier:
    if not secret:
        raise ValueError("empty secret")
    try:
        window = parse_go_duration(route.replay_window) if route.replay_window else timedelta(0)
    except GoDurationError as e:
        raise ValueError(f"replay_window: {e}") from e
    return StripeVerifier(secret=secret.encode("utf-8"), replay_window=window)


@dataclass(frozen=True, slots=True)
class StripeVerifier:
    secret: bytes
    replay_window: timedelta = timedelta(0)

    def verify(self, raw_body: bytes, headers, now: datetime) -> None:
        header = headers.get(HEADER)
        if not header:
            raise VerificationError(f"missing {HEADER} header")

        timestamp, signatures = _parse_header(header)
        if not timestamp or not signatures:
            raise VerificationError(f"malformed {HEADER} header")

        try:
            ts = int(timestamp)
        except ValueError:
            raise VerificationError("invalid timestamp") from None

        if self.replay_window > timedelta(0):
            delta = abs(now.timestamp() - ts)
            if delta > self.replay_window.total_seconds():
                raise VerificationError("timestamp outside replay window")

        # The signed payload is the timestamp, a literal dot, then the exact
        # bytes received -- concatenated, not formatted, so a body that is not
        # valid UTF-8 still hashes correctly.
        expected = hmac.new(
            self.secret,
            timestamp.encode("ascii") + b"." + raw_body,
            hashlib.sha256,
        ).digest()

        # Stripe may send several v1 signatures during a secret rotation; any
        # one matching is a pass. An unparseable candidate is skipped rather
        # than failing the request, matching Stripe's own library.
        for candidate in signatures:
            try:
                got = decode_hex(candidate)
            except ValueError:
                continue
            if hmac.compare_digest(got, expected):
                return
        raise VerificationError("no matching signature")


def _parse_header(header: str) -> tuple[str, list[str]]:
    """Split ``t=...,v1=...,v1=...`` into its timestamp and signatures.

    Unknown keys (Stripe has shipped ``v0`` alongside ``v1``) are ignored
    rather than treated as an error.
    """
    timestamp = ""
    signatures: list[str] = []
    for part in header.split(","):
        key, sep, value = part.strip().partition("=")
        if not sep:
            continue
        if key == "t":
            timestamp = value
        elif key == "v1":
            signatures.append(value)
    return timestamp, signatures
